Benjamin - Benjamin-Tu/bt_week4_assignment.do

```diff
@@ -0,0 +1,188 @@
+
+replace pixel_village=1 if tag_pixel_village_mean >1
+
+* c)
+gen village_categories=1 if pixel_village==0 
+
+replace village_categories=2 if pixel_village==1 & pixel_consistent==0
+
+replace village_categories=3 if pixel_village==1 & pixel_consistent==1
+
+tab hhid if village_categories==2
+
+*Q2
+use "$wd/psle_student_raw.dta", clear
+
+	levelsof schoolcode, local (list) // assign a number to each school
+	local n = 0 
+	foreach i of local list {
+		local n = `n' + 1 
+		use "$wd/psle_student_raw.dta", clear
+		keep if schoolcode == "`i'"
+
+		*Above mentioned parts are not relevant as you already have the webscraped info in the dataset.
+
+		split s, parse(">SUBJECTS") //we need to get rid of part before "Subject", I figured this out by eyeballing the data
+
+		*right now all the data is in the same row,Objective: each row is a unique student
+
+		*Identify any pattern => new line
+		*eyeball the data/website. There's a new line after subject grades. 
+		*Failry certain "</TD></TR>" is where a new line starts.
+		split s2, parse("</TD></TR>") gen(var) //
+		*Still every student info is in the same row,we can get to 1 row = 1 student info
+		*by using reshape command
+
+		gen serial = _n
+		reshape long var, i(serial) j(j)
+		split var, parse("</FONT></TD>")
+		*keep only the relevant variables
+		keep var1 var2 var3 var4 var5
+
+		*dropping first and last rows as they are empty
+		drop if var2=="" & var3==""
+
+
+		*candidate ID variable
+		gen cand_id = substr(var1,-14,.)
+
+		*gender
+		gen gender = substr(var3,-1,.)
+
+
+		*Prem Number
+		gen prem_number =  substr(var2,strpos(var2, "CENTER") +8, .)
+
+		*Name
+		gen name =  substr(var4,strpos(var4, "<P>") +3 , .)
+
+		*grades
+		replace var5 = substr(var5,ustrpos(var5, "LEFT") +6 , .)
+		replace var5 = substr(var5,1 , strlen(var5) - 7)
+
+		*all the subject info is in one columns, create separate columns
+		split var5, parse(,) //use "comma" as the parser.
+		*rename variables
+		rename var51 kiswahili
+		rename var52 english
+		rename var53 maarifa
+		rename var54 hisabati
+		rename var55 science
+		rename var56 uraia
+		rename var57 average
+
+		*drop columns that are no longer needed.
+		drop var1 var2 var3 var4 var5 
+
+		*extract just the greade
+		local varlist "kiswahili english maarifa hisabati science uraia average"
+		foreach subject in `varlist'{
+			replace `subject' = usubstr(`subject',-1,1)
+		}
+		
+		// generate variable for schoolcode
+		local y = "`i'" 
+		gen schoolcode = "`y'"
+		order schoolcode
+		
+		if `i' == 1 {
+			tempfile one
+			save `one', replace
+		}
+		else {
+			append using `one'
+			save `one', replace
+		}
+		
+		
+		
+	}
+
+Q3
+global excel_t21 "$wd/Pakistan_district_table21.xlsx"
+*update the global
+
+clear
+*setting up an empty tempfile
+tempfile table21
+save `table21', replace emptyok
+
+*Run a loop through all the excel sheets (135) this will take 2-10 mins because it has to import all 135 sheets, one by one
+forvalues i=1/135 {
```

Comment:
The final file here had some inconsistencies in the values of `table21` (e.g., some are "18 AND ABOVE", others are "OVERALL", others are "18 OR ABOVE"); I'm pretty sure all the values of the table21 variable should be the same.

Also, there are some blanks in the data, I think because your loop does not take into account that the variables are in different columns between the different Excel tables (which means that you need to find a way to remove or otherwise skip the empty columns as you read in the data).

It does look like this loop produces the right number of observations!